spec/integration/mentioning_spec.rb

```diff
@@ -22,74 +35,275 @@ def mention_stream_for(user)
+    end
 
+    it "receiver can see status message in streams" do
+      expect(status_msg).to be_in_streams_of(@user3)
+    end
+
+    it "non-receiver is not mentioned in status message" do
+      expect(@user2).not_to be_mentioned_in(status_msg)
+    end
+
+    it "non-receiver can't see status message in streams" do
+      expect(status_msg).not_to be_in_streams_of(@user2)
+    end
+  end
+
+  context "in private post via federation multiple targets" do
+    let(:status_msg) {
+      post_status_message_via_federation(text_mentioning(eve) + text_mentioning(bob), eve, bob)
+    }
+
+    [eve, bob].each do |user|
+      it "#{user.username} is mentioned in status message" do
+        expect(user).to be_mentioned_in(status_msg)
+      end
+
+      it "#{user.username} can see status message in streams" do
+        expect(status_msg).to be_in_streams_of(user)
+      end
+    end
+  end
+
+  it "mentions people in public posts" do
     status_msg = nil
     expect {
-      post "/status_messages.json", status_message: {text: text_mentioning(@user3)}, aspect_ids: "public"
-      status_msg = StatusMessage.find(JSON.parse(response.body)["id"])
+      status_msg = post_status_message(@user3, "public")
     }.to change(Post, :count).by(1)
 
     expect(status_msg).not_to be_nil
     expect(status_msg.public?).to be true
     expect(status_msg.text).to include(@user3.name)
     expect(status_msg.text).to include(@user3.diaspora_handle)
 
-    expect(stream_for(@user3).map(&:id)).to include(status_msg.id)
-    expect(mention_stream_for(@user3).map(&:id)).to include(status_msg.id)
+    expect(@user3).to be_mentioned_in(status_msg)
+    expect(status_msg).to be_in_streams_of(@user3)
   end
 
   it "mentions people that are in the target aspect" do
-    expect(users_connected?(@user1, @user2)).to be true
-
     status_msg = nil
     expect {
-      post "/status_messages.json", status_message: {text: text_mentioning(@user2)}, aspect_ids: default_aspect.id.to_s
-      status_msg = StatusMessage.find(JSON.parse(response.body)["id"])
+      status_msg = post_status_message(@user2, default_aspect.id.to_s)
     }.to change(Post, :count).by(1).and change(AspectVisibility, :count).by(1)
 
     expect(status_msg).not_to be_nil
     expect(status_msg.public?).to be false
     expect(status_msg.text).to include(@user2.name)
     expect(status_msg.text).to include(@user2.diaspora_handle)
 
-    expect(stream_for(@user2).map(&:id)).to include(status_msg.id)
-    expect(mention_stream_for(@user2).map(&:id)).to include(status_msg.id)
+    expect(@user2).to be_mentioned_in(status_msg)
+    expect(status_msg).to be_in_streams_of(@user2)
+  end
+
+  context "in comments" do
+    context "with public post" do
+      let(:status_msg) { FactoryGirl.create(:status_message, author: alice.person, public: true) }
+
+      context "a commenter may mention a person" do
+        [[bob, eve], [eve, bob], [alice, bob], [bob, alice], [local_luke, local_leia], [alice, eve]]
```

Comment:
Yes, these all are different type of relationships between users. I'll
check once again, but we need all of them because receive works
different depending on the relationship type. So it's easy to screw it up.